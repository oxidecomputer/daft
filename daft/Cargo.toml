[package]
name = "daft"
version = "0.1.0"
description = "Structural diffs of Rust data structures"
edition.workspace = true
repository.workspace = true
rust-version.workspace = true
license.workspace = true

[lints]
workspace = true

[dependencies]
daft-derive.workspace = true
newtype-uuid = { workspace = true, optional = true }
oxnet = { workspace = true, optional = true }
paste.workspace = true
uuid = { workspace = true, optional = true, features = ["v4"] }

[features]
newtype-uuid1 = ["dep:newtype-uuid"]
oxnet01 = ["dep:oxnet"]
uuid1 = ["dep:uuid"]
