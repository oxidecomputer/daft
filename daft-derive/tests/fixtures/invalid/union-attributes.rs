use daft::Diffable;

#[derive(Diffable)]
#[daft(leaf)]
union MyUnion {
    #[daft(leaf)]
    a: i32,
    #[daft(ignore)]
    b: u32,
}

#[derive(Diffable)]
#[daft(ignore)]
union MyUnion2 {
    a: i32,
    b: u32,
}

fn main() {
    // MyEnum/MyEnum2 and its Diffable impl should all exist.
    let before = MyUnion { a: 0 };
    let after = MyUnion { b: 1 };

    let _diff = before.diff(&after);

    let before = MyUnion2 { a: 0 };
    let after = MyUnion2 { a: 1 };

    let _diff = before.diff(&after);
}
