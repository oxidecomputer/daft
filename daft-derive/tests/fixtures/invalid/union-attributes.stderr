error: this is unnecessary: the Diffable implementation for unions is always a leaf
 --> tests/fixtures/invalid/union-attributes.rs:4:8
  |
4 | #[daft(leaf)]
  |        ^^^^

error: daft attributes are not allowed on union fields
 --> tests/fixtures/invalid/union-attributes.rs:6:5
  |
6 |     #[daft(leaf)]
  |     ^^^^^^^^^^^^^

error: daft attributes are not allowed on union fields
 --> tests/fixtures/invalid/union-attributes.rs:8:5
  |
8 |     #[daft(ignore)]
  |     ^^^^^^^^^^^^^^^

error: daft attributes are not allowed on unions
  --> tests/fixtures/invalid/union-attributes.rs:13:8
   |
13 | #[daft(ignore)]
   |        ^^^^^^
